## Homework 2 Problem 5 (b)

In [1]:
import numpy as np

### Jacobi method

In [2]:
def Jacobi(D, R, b, x_0, n):
    """
        D: diagonal matrix with dimension dim * dim
        R: remainder matrix, with dimension dim * dim
        A = D + R
        x: solution of previous iteration
        n: max number of iteration
    """
    dim, _ = D.shape
    x_current = x_0
    for i in range(n):
        print(i)
        x_next = []
        for j in range(dim):
            x_next.append(1.0 / D[j][j] * (b[j] - R[j].dot(x_current)))
        x_current = np.array(x_next).reshape(dim, 1)
        print(x_current)
    return x_current

In [3]:
D = np.array([[5, 0, 0], [0, 9, 0], [0, 0, -7]])
R = np.array([[0, -2, 3], [-3, 0, 1], [2, -2, 0]])
b = np.array([-1, 2, 3]).reshape(3, 1)
x_0 = np.array([0, 0, 0]).reshape(3, 1)

In [4]:
Jacobi(D, R, b, x_0, 10)

0
[[-0.2       ]
 [ 0.22222222]
 [-0.42857143]]
1
[[ 0.14603175]
 [ 0.2031746 ]
 [-0.54920635]]
2
[[ 0.21079365]
 [ 0.3319224 ]
 [-0.44489796]]
3
[[ 0.19970773]
 [ 0.34191988]
 [-0.46317964]]
4
[[ 0.21467574]
 [ 0.34025587]
 [-0.46920347]]
5
[[ 0.21762443]
 [ 0.34591452]
 [-0.46445147]]
6
[[ 0.21703669]
 [ 0.34636942]
 [-0.46522574]]
7
[[ 0.21768321]
 [ 0.34625953]
 [-0.46552364]]
8
[[ 0.217818  ]
 [ 0.34650814]
 [-0.46530752]]
9
[[ 0.21778777]
 [ 0.34652906]
 [-0.46534004]]


array([[ 0.21778777],
       [ 0.34652906],
       [-0.46534004]])

Therefore, after ten steps, the answer is $[0.217788 \ \ 0.346529 \ \ -0.465340]^T$.

### Gauss-Seidel method

In [5]:
def Gauss_Seidel(D, R, b, x_0, n):
    """
        D: diagonal matrix with dimension dim * dim
        R: remainder matrix, with dimension dim * dim
        A = D + R
        x: solution of previous iteration
        n: max number of iteration
    """
    dim, _ = D.shape
    x_current = x_0
    for i in range(n):
        print(i)
        for j in range(dim):
            x_current[j] = 1.0 / D[j][j] * (b[j] - R[j].dot(x_current))
            print(x_current[j])
    return x_current

In [6]:
D = np.array([[5, 0, 0], [0, 9, 0], [0, 0, -7]])
R = np.array([[0, -2, 3], [-3, 0, 1], [2, -2, 0]])
b = np.array([-1, 2, 3]).reshape(3, 1)
x_0 = np.array([0., 0., 0.]).reshape(3, 1)

In [7]:
Gauss_Seidel(D, R, b, x_0, 10)

0
[-0.2]
[0.15555556]
[-0.53015873]
1
[0.18031746]
[0.34123457]
[-0.47454775]
2
[0.22122247]
[0.34869057]
[-0.46499089]
3
[0.21847076]
[0.34671146]
[-0.46521163]
4
[0.21781156]
[0.34651626]
[-0.4653442]
5
[0.21781302]
[0.34653147]
[-0.46534813]
6
[0.21782147]
[0.34653473]
[-0.46534665]
7
[0.21782188]
[0.3465347]
[-0.46534652]
8
[0.21782179]
[0.34653465]
[-0.46534653]
9
[0.21782178]
[0.34653465]
[-0.46534653]


array([[ 0.21782178],
       [ 0.34653465],
       [-0.46534653]])

For Gauss-Seidel method, after ten steps, the answer is $[0.217822 \ \ 0.346535 \ \ -0.465347]^T$.